pkg/cmd/infra/router/template.go

```diff
@@ -565,3 +566,77 @@ func (o *TemplateRouterOptions) blueprintRoutes(routeclient *routeclientset.Clie
 
 	return blueprints, nil
 }
+
+// makeTLSConfig checks whether metrics TLS is configured and
+// if so returns a tls.Config whose certificate is automatically
+// reloaded on the given period.
+func makeTLSConfig(reloadPeriod time.Duration) (*tls.Config, error) {
+	certFile := env("ROUTER_METRICS_TLS_CERT_FILE", "")
+	if len(certFile) == 0 {
+		return nil, nil
+	}
+	keyFile := env("ROUTER_METRICS_TLS_KEY_FILE", "")
+
+	// Load the initial certificate contents.
+	certBytes, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return nil, err
+	}
+	keyBytes, err := ioutil.ReadFile(keyFile)
+	if err != nil {
+		return nil, err
+	}
+	certificate, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Safely reload the certificate on a fixed period for simplicity.
+	ticker := time.NewTicker(reloadPeriod)
+	var lock sync.Mutex
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				// Check for certificate or key content changes.
+				latestCertBytes, err := ioutil.ReadFile(certFile)
+				if err != nil {
+					log.Error(err, "failed to read certificate file", "cert", certFile, "key", keyFile)
+					break
+				}
+				latestKeyBytes, err := ioutil.ReadFile(keyFile)
+				if err != nil {
+					log.Error(err, "failed to read key file", "cert", certFile, "key", keyFile)
+					break
+				}
+
+				if bytes.Equal(latestCertBytes, certBytes) && bytes.Equal(latestKeyBytes, keyBytes) {
+					// Nothing changed, try later.
+					break
+				}
+
+				// Something changed, refresh the certificate.
+				latest, err := tls.X509KeyPair(latestCertBytes, latestKeyBytes)
+				if err != nil {
+					log.Error(err, "failed to reload certificate", "cert", certFile, "key", keyFile)
+					break
+				}
+				lock.Lock()
+				certificate = latest
+				certBytes = latestCertBytes
+				keyBytes = latestKeyBytes
+				lock.Unlock()
```

Comment:
We only need the lock around `certificate = latest`, right?  That is, we don't need the lock when assigning `latestCertBytes` and `latestKeyBytes`?  It's confusing to lock during access of `latestCertBytes` and `latestKeyBytes` here but not for the comparison above.

Reply:
You're right — `certBytes` and `keyBytes` are single reader/writer and don't require guarding.

Fixed.